src/Parser.hs

```diff
@@ -1,16 +1,93 @@
 {-# LANGUAGE OverloadedStrings, GADTs #-}
-module Parser (parseCmd) where
+module Parser (parseCmd, doPreprocess) where
 
 import Control.Monad
 import Control.Monad.IO.Class
+import Control.Monad.Combinators.Expr
 import Data.Char
 import qualified Data.Text as T
 import Data.Text (Text)
+import Data.Maybe
 import Text.Megaparsec
 import Text.Megaparsec.Char
 import qualified Text.Megaparsec.Char.Lexer as L
+
 import Abs
 import Builtins
+import Utils
+
+-- | Preprocessing part.
+
+doPreprocess :: String -> Shell String
+doPreprocess t = do
+  ecpeb <- runParserT (linePreprocessor 0) "preprocessing" $ T.pack t
+  either (\peb -> t <$ liftIO (putStrLn $ errorBundlePretty peb)) return ecpeb
+
+linePreprocessor :: Integer -> Parser String
+linePreprocessor acc = do
+  x <- optional pCharAndEscape
+  case x of
+    Nothing -> return ""
+    Just x' -> do
+      let escape = x' == '\\'
+      let acc' = if escape then acc + 1 else 0
+      if x' == '$' && acc `mod` 2 == 0 
+        then do
+          val <- preprocessVar  
+          next <- linePreprocessor acc'
+          return $ val ++ next
+        else do
+          next <- linePreprocessor acc'
+          return $ (if escape && acc' `mod` 2 == 1 then "" else [x']) ++ next
+
+preprocessVar :: Parser String
+preprocessVar = do
+    var <- pName
+    val <- getVar var
+    return $ show $ fromMaybe (VStr "") val
```

Comment:
Note that `show (VStr "val") == "VStr \"val\""`. This should either be a custom printing function, or a custom Show instance (not `deriving`).